akka-stream-testkit/src/test/scala/akka/stream/testkit/Utils.scala

```diff
@@ -25,4 +32,67 @@ object Utils {
     case _ =>
       throw new Exception(s"Unable to determine dispatcher of $ref")
   }
+
+  /**
+   * INTERNAL API
+   *
+   * Returns a flow that is almost identity but delays propagation of cancellation from downstream to upstream.
+   *
+   * Borrowed from https://github.com/akka/akka-http/blob/d85b71df0f01ce80f65fbb17e513fc2beb6d5adf/akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala#L193
+   */
+  def delayCancellation[T](cancelAfter: Duration): Flow[T, T, NotUsed] =
+    Flow.fromGraph(new DelayCancellationStage(cancelAfter))
+  final class DelayCancellationStage[T](cancelAfter: Duration) extends SimpleLinearGraphStage[T] {
+    def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+      new GraphStageLogic(shape) with ScheduleSupport with InHandler with OutHandler with StageLogging {
+        setHandlers(in, out, this)
+
+        def onPush(): Unit =
+          push(out, grab(in)) // using `passAlong` was considered but it seems to need some boilerplate to make it work
+        def onPull(): Unit = pull(in)
+
+        var timeout: OptionVal[Cancellable] = OptionVal.None
+
+        override def onDownstreamFinish(): Unit = {
+          cancelAfter match {
+            case finite: FiniteDuration =>
+              log.debug(s"Delaying cancellation for $finite")
+              timeout = OptionVal.Some {
+                scheduleOnce(finite) {
+                  log.debug(s"Stage was canceled after delay of $cancelAfter")
+                  timeout = OptionVal.None
+                  completeStage()
+                }
+              }
+            case _ => // do nothing
+          }
+
+          // don't pass cancellation to upstream but keep pulling until we get completion or failure
+          setHandler(in, new InHandler {
+            if (!hasBeenPulled(in)) pull(in)
+
+            def onPush(): Unit = {
+              grab(in) // ignore further elements
+              pull(in)
+            }
+          })
+        }
+
+        override def postStop(): Unit = timeout match {
+          case OptionVal.Some(x) => x.cancel()
+          case OptionVal.None    => // do nothing
+        }
+      }
+  }
+
+  trait ScheduleSupport { self: GraphStageLogic =>
```

Comment:
internal api? but wouldn't it better to use `TimerGraphStageLogic` instead of introducing another way